### ISSUE TITLE
`unsound-return-statement` false positive on `join(sorted(..., key=lambda))`

### ISSUE THREAD

**Issue by mxr** [astral-sh/ty#4311]
### Summary

`ty` fails to infer the type of a `sorted(..., key=<unannotated lambda>)` call when it's nested under `join`. Not sure if related to #3597. `mypy` handles this correctly.

Repro with counterexamples:

```python
items: list[str] = ["b", "a"]

# bug: nested lambda key -> revealed as Unknown
joined = ",".join(sorted(items, key=lambda i: i))
reveal_type(joined)  # Unknown, expected str

# counterexample: key=len -> revealed as str
joined = ",".join(sorted(items, key=len))
reveal_type(joined)  # str

# counterexample: no key -> revealed as str
joined = ",".join(sorted(items))
reveal_type(joined)  # str

# counterexample: split across separate statements -> revealed as str
s = sorted(items, key=lambda i: i)
joined = ",".join(s)
reveal_type(joined)  # str
```

- [Mypy playground (only with the first stanza)](https://mypy-play.net/?gist=2160f6c63ae4685a72c0c3ddb13fbdf3)
- [Ty playground](https://play.ty.dev/8117f30e-a6f9-4c70-b1df-defd7736f596)

### Version

ty 0.0.72 (62ebbecc7 2026-08-14)

**Comment by carljm** [comment]
Thanks! Definitely a bug in generic inference. This is not specific to `sorted`, string comparison, or typeshed. Here is a smaller reproduction, with all types surfaced:

```python
from collections.abc import Callable
from typing import overload, reveal_type

class Parent: ...
class Child(Parent): ...

@overload
def inner[T](value: T, key: None = None) -> T: ...
@overload
def inner[T](value: T, key: Callable[[T], object]) -> T: ...
def inner[T](value: T, key: Callable[[T], object] | None = None) -> T:
    return value

@overload
def outer(value: Child) -> Child: ...
@overload
def outer(value: Parent) -> Parent: ...
def outer(value: Parent) -> Parent:
    return value

def f(value: Parent):
    reveal_type(inner(value, lambda _: None))         # Parent
    reveal_type(outer(inner(value, lambda _: None)))  # Unknown; expected Parent
```

It looks like the failure is in speculative contextual inference. In the original example, `str.join` tries its `Iterable[LiteralString]` overload, which gives `sorted` an expected return type that conflicts with the actual `list[str]` argument. We eagerly use that context to specialize the lambda’s parameter. The inner overloaded call then fails and recovers as `Unknown`. That recovery type keeps the outer `LiteralString` overload alive, making the outer call appear ambiguous.

The relevant code is [`argument_type_context_specialization`](https://github.com/astral-sh/ruff/blob/ce46a36be796faff60d1ad5734e571bb1223c0a7/crates/ty_python_semantic/src/types/call/bind.rs#L7652-L7728), followed by the [failed-overloaded-call return fallback](https://github.com/astral-sh/ruff/blob/ce46a36be796faff60d1ad5734e571bb1223c0a7/crates/ty_python_semantic/src/types/call/bind.rs#L4411-L4435).

This is related to #3559 and astral-sh/ruff#26680, but that PR doesn't fix it. The PR improves handling of return-context constraints, but its argument-context specialization still retains the declared context when the argument-derived solution is incompatible with it. I think the work in #3933 and #3934 is also relevant here.

**Comment by ibraheemdev** [comment]
I'm not sure this is actually related to bidirectional inference. While the eager specialization can lead to false positives, in this case the first overload arm is actually unsatisfiable. The actual problem seems to be the `Unknown` recovery type, which makes the overload look ambiguous despite having a single satisfying arm. Maybe the fix here is to directly filter overloads with unsatisfiable call arguments during overload resolution?

**Comment by carljm** [comment]
Yeah, I think your analysis is correct. I'll remove that tag.
